### PR TITLE
Share file with other user and delete shared with users

### PR DESCRIPTION
**Pull request by bishalrijal**
## Description
Test for deleting shared with user using webUI 

## How Has This Been Tested?
locally 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [X] Tests only (no source changes)

## Checklist:
- [ ] Code changes
- [ ] Unit tests added
- [X] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
- [X] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

File: tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature [hunk 1]
```diff
@@ -225,3 +225,10 @@ Feature: Sharing files and folders with internal users
     #And the share-with field should not be visible in the details panel
     Then the share-with field should be visible in the details panel
     And user "user1" should not be able to share file "testimage (2).jpg" with user "User Three" using the sharing API
+
+  Scenario: user shares the file/folder with another internal user and delete the share with user
+    Given user "user1" has logged in using the webUI
+    And the user shares file "lorem.txt" with user "User Two" using the webUI
+    When the user delete share with user "User Two"
```
Comment (dpakach) [inline tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature]: Delete share for which file? It is not clear. 
I think it should be 
`
    When the user delete share with user "User Two" for the current file
`

Comment (dpakach) [reply tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature]: Same goes for the next step

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 2]
```diff
@@ -571,6 +600,26 @@ public function checkPublicLinkCount(Session $session, $count) {
 			throw new \Exception("Found $publicLinkTitlesCount public link entries but expected $count");
 		}
 	}
+
+	/**
+	 * check if user with the given username is present
```
Comment (dpakach) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]:
```suggestion
	 * check if user with the given username is present in the shared with list
```

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 3]
```diff
@@ -571,6 +600,26 @@ public function checkPublicLinkCount(Session $session, $count) {
...
+	 *
+	 * @param Session $session
+	 * @param string $username
+	 *
+	 * @return bool
+	 */
+	public function isUserPresentInShareWithList(Session $session, $username) {
+		$this->waitForAjaxCallsToStartAndFinish($session);
```
Comment (dpakach) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]: Is this required here?

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 4]
```diff
@@ -556,6 +559,32 @@ public function checkThatNameIsNotInPublicLinkList(Session $session, $entryName)
...
+	 * @param Session $session
+	 * @param string $username
+	 *
+	 * @return void
+	 */
+	public function deleteShareWithUser(Session $session, $username) {
+		$shareWithList = $this->getShareWithList();
+		foreach ($shareWithList as $userOrGroup) {
+			if ($userOrGroup->find('xpath', $this->userNameSpanXpath)->getHtml() === $username) {
+				$userOrGroup->find('xpath', "//a[contains(@class,'unshare')]")->click();
```
Comment (dpakach) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]: Do not use xpath here. make a variable.

File: tests/acceptance/features/bootstrap/WebUISharingContext.php [hunk 5]
```diff
@@ -230,6 +240,17 @@ public function theUserHasOpenedThePublicLinkShareTab() {
...
 	}
+	/**
+	 * @When the user delete share with user :username
+	 *
+	 * @param string $username
+	 *
+	 * @return void
+	 */
+	public function theUserDeleteShareWithUser($username) {
+		$this->sharingDialog = $this->filesPage->openSharingDialog("lorem.txt", $this->getSession());
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUISharingContext.php]: Why does it use explicit file name (i.e. "lorem.txt")?
Can we not do for other files?

File: tests/acceptance/features/bootstrap/WebUISharingContext.php [hunk 6]
```diff
@@ -905,6 +926,30 @@ public function theAutocompleteListShouldNotBeDisplayed() {
...
 
+	/**
+	 * @Then the user :username should not be in share with user list
+	 *
+	 * @param string $username
+	 *
+	 * @return void
+	 */
+	public function theUserShouldNotBeInShareWithUserList($username) {
+		if ($this->sharingDialog->isUserPresentInShareWithList($this->getSession(), $username)) {
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUISharingContext.php]: Use assertion here from PHPUnitFramework

File: tests/acceptance/features/bootstrap/WebUISharingContext.php [hunk 7]
```diff
@@ -905,6 +926,30 @@ public function theAutocompleteListShouldNotBeDisplayed() {
...
+	/**
+	 * @Then the file :fileName should not be in share with other list
+	 *
+	 * @param string $fileName
+	 *
+	 * @return void
+	 */
+	public function theFileShouldNotBeInShareWithOtherList($fileName) {
+		$this->webUIFilesContext->theUserBrowsesToTheSharedWithOthersPage();
+		$this->sharedWithOthersPage->isEmpty();
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUISharingContext.php]: There might be still other files in the `SharedWithOthersPage`

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php (hunk 4 above)
Comment (dpakach) [reply tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]: Also make sure to wait for any ajax calls

**Review by dpakach — CHANGES_REQUESTED** [review]

File: tests/acceptance/features/bootstrap/WebUISharingContext.php [hunk 8]
```diff
@@ -100,6 +106,7 @@ class WebUISharingContext extends RawMinkContext implements Context {
 	private $oldFedSharingFallbackSetting = null;
 
 	private $publicShareTab;
+	private $userAndGroupsTab;
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUISharingContext.php]: I dont think this variable is used anywhere.

File: tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature [hunk 9]
```diff
@@ -225,3 +225,10 @@ Feature: Sharing files and folders with internal users
...
+    Then the user "User Two" should not be in share with user list
+    And the file "lorem.txt" should not be in share with other list
```
Comment (dpakach) [inline tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature]:
```suggestion
    And the file "lorem.txt" should not be in shared with other list
```

**Review by dpakach — DISMISSED** [review]

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 10]
```diff
@@ -556,6 +559,32 @@ public function checkThatNameIsNotInPublicLinkList(Session $session, $entryName)
 		}
 	}
 
+	/**
+	 * get the list of shared with list
```
Comment (dpakach) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]:
```suggestion
	 * get the list of shared with users
```

File: tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature [hunk 11]
```diff
@@ -225,3 +225,11 @@ Feature: Sharing files and folders with internal users
...
     Then the share-with field should be visible in the details panel
     And user "user1" should not be able to share file "testimage (2).jpg" with user "User Three" using the sharing API
+
+  Scenario: user shares the file/folder with another internal user and delete the share with user
+    Given user "user1" has logged in using the webUI
+    And user "user1" has shared file "lorem.txt" with user "user2"
+    When the user opens the share dialog for file "lorem.txt"
+    And the user deletes share with user "User Two" for the current file
+    Then the user "User Two" should not be in share with user list
+    And file "lorem.txt" should not be listed in shared-with-others page on the webUI
```
Comment (paurakhsharma) [inline tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature]: Maybe you can also add `And as "user2" file "lorem (2).txt" should not exist`

**Comment by individual-it** [comment]
@ownclouders rebase

**Comment by ownclouders** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by ownclouders** [comment]
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

**Review by skshetry — COMMENTED** [review]
Looks good to me. I'll add phil-davis and individual-it to the reviewers list.

@paurakhsharma please review again
@dpakach please review and discard your earlier request for changes.

**Review by individual-it — APPROVED** [review]

**Comment by individual-it** [comment]
not part of the sprint, as its a project for trainees

**Pull request merged**